packages/angular_devkit/build_facade/src/build-system.ts

```diff
@@ -0,0 +1,98 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { Path, asPosixPath, asWindowsPath, dirname, logging, schema } from '@angular-devkit/core';
+import { readFileSync } from 'fs';
+import { Observable } from 'rxjs/Observable';
+import { of } from 'rxjs/observable/of';
+import { _throw } from 'rxjs/observable/throw';
+import { mergeMap } from 'rxjs/operators';
+import { TargetOptions } from './architect';
+
+
+export interface BuildContext {
+  logger: logging.Logger;
+}
+
+export interface BuildStats {
+  start: Date;
+  end: Date;
+  hash: string;
+
+  totalSize: number;
+  criticalSize: number;
+  initialSize: number;
+  assetsSize: number;
+}
+
+export interface BuildAsset {
+  // TODO: figure out virtual file system host
+  // host: Host;
+  name: string;
+  path: Path;
+  size: number;
+  flags: string[];
+}
+
+export interface BuildArtifact {
+  name: string;
+  outputs: BuildAsset[];
+  flags: string[];
+}
+
+export interface BuildEvent {
```

Comment:
Look into the `BuildEventProtocol` from Bazel, and lets use that interface. Then the mapping will be easy and it works with any type system (and is extensible).

https://docs.bazel.build/versions/master/build-event-protocol.html

You can find the protos here: https://github.com/googleapis/googleapis/tree/master/google/devtools/build/v1

Can be done in a separate PR but this is required for serve since it relies on build.

Reply:
I agree this is how we should transmit build artifacts but at the moment `serve` needs a bit more than this. It actually needs the webpack compilation since it hooks into it and adds entry points for live reload.

e2e only needs to know when serve has emitted a build event, and doesn't care about the content.

I think what we want right now is to hook build and serve directly while we figure out the build events and dependency model. When we have that, maybe rework the dev server to be more generic.